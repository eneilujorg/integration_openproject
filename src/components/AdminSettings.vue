<template>
	<div id="openproject_prefs" class="section">
		<SettingsTitle />
		<div class="openproject-server-host">
			<FormHeading index="1"
				:title="t('integration_openproject', 'OpenProject Server')"
				:is-complete="isServerHostFormComplete" />
			<FieldValue v-if="isServerHostFormInView"
				is-required
				class="pb-1"
				:title="t('integration_openproject', 'OpenProject host')"
				:value="state.oauth_instance_url" />
			<TextInput v-else
				id="openproject-oauth-instance"
				ref="openproject-oauth-instance-input"
				v-model="serverHostUrlForEdit"
				is-required
				:read-only="isServerHostUrlReadOnly"
				class="pb-2"
				:label="t('integration_openproject', 'OpenProject host')"
				place-holder="https://www.my-openproject.com"
				:hint-text="t('integration_openproject', 'Please introduce your OpenProject host name')"
				:error-message="serverHostErrorMessage"
				@click="isServerHostUrlReadOnly = false" />
			<div class="d-flex">
				<Button v-if="isServerHostFormInView"
					data-test-id="reset-server-host-btn"
					icon-class="pencil-icon"
					:text="t('integration_openproject', 'Edit server information')"
					@click="setServerHostFormToEditMode" />
				<Button v-if="isServerHostFormComplete && isServerHostFormInEdit"
					:text="t('integration_openproject', 'Cancel')"
					data-test-id="cancel-edit-server-host-btn"
					@click="setServerHostFormToViewMode" />
				<Button v-if="isServerHostFormInEdit"
					class="submit-btn"
					icon-class="check-icon"
					:text="t('integration_openproject', 'Save')"
					:is-loading="loadingServerHostForm"
					:is-disabled="!serverHostUrlForEdit || serverHostUrlForEdit === state.oauth_instance_url"
					@click="saveOpenProjectHostUrl" />
			</div>
		</div>
		<div class="openproject-oauth-values">
			<FormHeading index="2"
				:title="t('integration_openproject', 'OpenProject OAuth settings')"
				:is-complete="isOPOAuthFormComplete"
				:is-disabled="isOPOAuthFormInDisableMode" />
			<div v-if="isServerHostFormComplete">
				<FieldValue v-if="isOPOAuthFormInView"
					is-required
					:value="state.client_id"
					:title="t('integration_openproject', 'OpenProject OAuth client ID')" />
				<TextInput v-else
					id="openproject-oauth-client-id"
					v-model="state.client_id"
					class="pb-2"
					is-required
					:label="t('integration_openproject', 'OpenProject OAuth client ID')"
					:hint-text="openProjectClientHint" />
				<FieldValue v-if="isOPOAuthFormInView"
					is-required
					class="pb-1"
					encrypt-value
					:title="t('integration_openproject', 'OpenProject OAuth client secret')"
					:value="state.client_secret" />
				<TextInput v-else
					id="openproject-oauth-client-secret"
					v-model="state.client_secret"
					is-required
					class="pb-2"
					:label="t('integration_openproject', 'OpenProject OAuth client secret')"
					:hint-text="openProjectClientHint" />
				<Button v-if="isOPOAuthFormComplete && isOPOAuthFormInView"
					data-test-id="reset-op-oauth-btn"
					icon-class="reset-icon"
					:text="t('integration_openproject', 'Reset OpenProject OAuth values')"
					@click="resetOPOAuthClientValues" />
				<Button v-else
					class="submit-btn"
					data-test-id="submit-op-oauth-btn"
					:text="t('integration_openproject', 'Save')"
					icon-class="check-icon"
					:is-loading="loadingOPOauthForm"
					:is-disabled="!state.client_id || !state.client_secret"
					@click="saveOPOAuthClientValues" />
			</div>
		</div>
		<div class="nextcloud-oauth-values">
			<FormHeading index="3"
				:title="t('integration_openproject', 'Nextcloud OAuth client')"
				:is-complete="isNcOAuthFormComplete"
				:is-disabled="isNcOAuthFormInDisableMode" />
			<div v-if="state.nc_oauth_client">
				<TextInput v-if="isNcOAuthFormInEdit"
					id="nextcloud-oauth-client-id"
					v-model="state.nc_oauth_client.clientId"
					class="pb-2"
					read-only
					is-required
					with-copy-btn
					:label="t('integration_openproject', 'Nextcloud OAuth client ID')"
					:hint-text="nextcloudClientHint" />
				<FieldValue v-else
					:title="t('integration_openproject', 'Nextcloud OAuth client ID')"
					:value="state.nc_oauth_client.clientId"
					is-required />
				<TextInput v-if="isNcOAuthFormInEdit"
					id="nextcloud-oauth-client-secret"
					v-model="state.nc_oauth_client.clientSecret"
					class="pb-2"
					read-only
					is-required
					with-copy-btn
					:label="t('integration_openproject', 'Nextcloud OAuth client secret')"
					:hint-text="nextcloudClientHint" />
				<FieldValue v-else
					:title="t('integration_openproject', 'Nextcloud OAuth client secret')"
					is-required
					encrypt-value
					with-inspection
					:value="ncClientSecret" />
				<Button v-if="isNcOAuthFormInEdit"
					class="submit-btn"
					:text="t('integration_openproject', 'Yes, I have copied these values')"
					icon-class="check-icon"
					:is-disabled="!ncClientId || !ncClientSecret"
					@click="setNCOAuthFormToViewMode" />
				<Button v-else
					data-test-id="reset-nc-oauth-btn"
					icon-class="reset-icon"
					:text="t('integration_openproject', 'Reset Nextcloud OAuth values')"
					@click="resetNcOauthValues" />
			</div>
		</div>
	</div>
</template>

<script>
import util from 'util'
import axios from '@nextcloud/axios'
import '@nextcloud/dialogs/styles/toast.scss'
import { loadState } from '@nextcloud/initial-state'
import { generateUrl } from '@nextcloud/router'
import { showSuccess, showError } from '@nextcloud/dialogs'
import Button from './admin/Button'
import TextInput from './admin/TextInput'
import FieldValue from './admin/FieldValue'
import FormHeading from './admin/FormHeading'
import SettingsTitle from '../components/settings/SettingsTitle'
import { F_MODES } from './../utils'

export default {
	name: 'AdminSettings',

	components: {
		Button,
		FieldValue,
		FormHeading,
		TextInput,
		SettingsTitle,
	},
	data() {
		return {
			formMode: {
				// server host form is never disabled.
				// it's either editable or view only
				server: F_MODES.EDIT,
				opOauth: F_MODES.DISABLE,
				ncOauth: F_MODES.DISABLE,
			},
			isFormCompleted: {
				server: false, opOauth: false, ncOauth: false,
			},
			loadingServerHostForm: false,
			loadingOPOauthForm: false,
			isOpenProjectInstanceValid: null,
			state: loadState('integration_openproject', 'admin-config'),
			isAdminConfigOk: loadState('integration_openproject', 'admin-config-status'),
			serverHostUrlForEdit: null,
			isServerHostUrlReadOnly: true,
		}
	},
	computed: {
		ncClientId() {
			return this.state.nc_oauth_client?.clientId
		},
		ncClientSecret() {
			return this.state.nc_oauth_client?.clientSecret
		},
		serverHostErrorMessage() {
			if (
				this.serverHostUrlForEdit === ''
				|| this.isOpenProjectInstanceValid === null
				|| this.isOpenProjectInstanceValid
			) return null
			return t('integration_openproject', 'Please introduce a valid OpenProject host name')
		},
		isServerHostFormComplete() {
			return this.isFormCompleted.server
		},
		isOPOAuthFormComplete() {
			return this.isFormCompleted.opOauth
		},
		isNcOAuthFormComplete() {
			return this.isFormCompleted.ncOauth
		},
		isServerHostFormInView() {
			return this.formMode.server === F_MODES.VIEW
		},
		isOPOAuthFormInView() {
			return this.formMode.opOauth === F_MODES.VIEW
		},
		isServerHostFormInEdit() {
			return this.formMode.server === F_MODES.EDIT
		},
		isNcOAuthFormInEdit() {
			return this.formMode.ncOauth === F_MODES.EDIT
		},
		isOPOAuthFormInDisableMode() {
			return this.formMode.opOauth === F_MODES.DISABLE
		},
		isNcOAuthFormInDisableMode() {
			return this.formMode.ncOauth === F_MODES.DISABLE
		},
		adminFileStorageHref() {
			let hostPart = ''
			const urlPart = '%sadmin/settings/storages'
			if (this.state?.oauth_instance_url.endsWith('/')) {
				hostPart = this.state.oauth_instance_url
			} else hostPart = this.state.oauth_instance_url + '/'
			return util.format(urlPart, hostPart)
		},
		openProjectClientHint() {
			const linkText = t('integration_openproject', 'Administration > File storages')
			const htmlLink = `<a class="link" href="${this.adminFileStorageHref}" target="_blank" title="${linkText}">${linkText}</a>`
			return t('integration_openproject', 'Go to your OpenProject {htmlLink} as an Administrator and start the setup and copy the values here.', { htmlLink }, null, { escape: false })
		},
		nextcloudClientHint() {
			const linkText = t('integration_openproject', 'Administration > File storages')
			const htmlLink = `<a class="link" href="${this.adminFileStorageHref}" target="_blank" title="${linkText}">${linkText}</a>`
			return t('integration_openproject', 'Copy the following values back into the OpenProject {htmlLink} as an Administrator.', { htmlLink }, null, { escape: false })
		},
	},
	created() {
		if (this.state) {
			if (this.state.oauth_instance_url) {
				this.formMode.server = F_MODES.VIEW
				this.isFormCompleted.server = true
			}
			if (!!this.state.client_id && !!this.state.client_secret) {
				this.formMode.opOauth = F_MODES.VIEW
				this.isFormCompleted.opOauth = true
			}
			if (this.state.oauth_instance_url) {
				if (!this.state.client_id || !this.state.client_secret) {
					this.formMode.opOauth = F_MODES.EDIT
				}
			}
			if (this.state.nc_oauth_client) {
				this.formMode.ncOauth = F_MODES.VIEW
				this.isFormCompleted.ncOauth = true
			}
		}
	},
	methods: {
		setServerHostFormToViewMode() {
			this.formMode.server = F_MODES.VIEW
		},
		setServerHostFormToEditMode() {
			this.formMode.server = F_MODES.EDIT
			// set the edit variable to the current saved value
			this.serverHostUrlForEdit = this.state.oauth_instance_url
			this.isOpenProjectInstanceValid = null
		},
		setNCOAuthFormToViewMode() {
			this.formMode.ncOauth = F_MODES.VIEW
			this.isFormCompleted.ncOauth = true
		},
		async saveOpenProjectHostUrl() {
			this.loadingServerHostForm = true
			await this.validateOpenProjectInstance()
			if (this.isOpenProjectInstanceValid) {
				const saved = await this.saveOPOptions()
				if (saved) {
					this.state.oauth_instance_url = this.serverHostUrlForEdit
					this.formMode.server = F_MODES.VIEW
					this.isFormCompleted.server = true
					if (!this.isFormCompleted.opOauth) {
						this.formMode.opOauth = F_MODES.EDIT
					}
				}
			}
			this.loadingServerHostForm = false
		},
		async saveOPOAuthClientValues() {
			await this.saveOPOptions()
			if (this.isAdminConfigOk) {
				this.formMode.opOauth = F_MODES.VIEW
				this.isFormCompleted.opOauth = true

				// if we do not have Nextcloud OAuth client yet, a new client is created
				if (!this.state.nc_oauth_client) {
					this.createNCOAuthClient()
				}
			}
		},
		resetOPOAuthClientValues() {
			OC.dialogs.confirmDestructive(
				t('integration_openproject', 'If you proceed you will need to update these settings with the new OpenProject OAuth credentials. Also, all users will need to reauthorize access to their OpenProject account.'),
				t('integration_openproject', 'Replace OpenProject OAuth values'),
				{
					type: OC.dialogs.YES_NO_BUTTONS,
					confirm: t('integration_openproject', 'Yes, replace'),
					confirmClasses: 'error',
					cancel: t('integration_openproject', 'Cancel'),
				},
				async (result) => {
					if (result) {
						await this.clearOPOAuthClientValues()
					}
				},
				true
			)
		},
		async clearOPOAuthClientValues() {
			this.formMode.opOauth = F_MODES.EDIT
			this.isFormCompleted.opOauth = false
			this.state.client_id = null
			this.state.client_secret = null
			const saved = await this.saveOPOptions()
			if (!saved) {
				this.formMode.opOauth = F_MODES.VIEW
				this.isFormCompleted.opOauth = true
			}
		},
		async validateOpenProjectInstance() {
			const url = generateUrl('/apps/integration_openproject/is-valid-op-instance')
			const response = await axios.post(url, { url: this.serverHostUrlForEdit })
			if (response.data !== true) {
				showError(
					t('integration_openproject', 'No OpenProject detected at the URL')
				)
				this.isOpenProjectInstanceValid = false
				await this.$nextTick()
				await this.$refs['openproject-oauth-instance-input']?.$refs?.textInput?.focus()
			} else {
				this.isOpenProjectInstanceValid = true
				this.state.oauth_instance_url = this.serverHostUrlForEdit
			}
		},
		async saveOPOptions() {
			const url = generateUrl('/apps/integration_openproject/admin-config')
			const req = {
				values: {
					client_id: this.state.client_id,
					client_secret: this.state.client_secret,
					oauth_instance_url: this.state.oauth_instance_url,
				},
			}
			try {
				const response = await axios.put(url, req)
				// after successfully saving the admin credentials, the admin config status needs to be updated
				this.isAdminConfigOk = response?.data?.status === true
				showSuccess(t('integration_openproject', 'OpenProject admin options saved'))
				return true
			} catch (error) {
				console.debug(error)
				showError(
					t('integration_openproject', 'Failed to save OpenProject admin options')
				)
				return false
			}
		},
		resetNcOauthValues() {
			OC.dialogs.confirmDestructive(
				t('integration_openproject', 'If you proceed you will need to update the settings in your OpenProject with the new Nextcloud OAuth credentials. Also, all users in OpenProject will need to reauthorize access to their Nextcloud account.'),
				t('integration_openproject', 'Replace Nextcloud OAuth values'),
				{
					type: OC.dialogs.YES_NO_BUTTONS,
					confirm: t('integration_openproject', 'Yes, replace'),
					confirmClasses: 'error',
					cancel: t('integration_openproject', 'Cancel'),
				},
				async (result) => {
					if (result) {
						this.state.nc_oauth_client = null
						this.createNCOAuthClient()
					}
				},
				true
			)
		},
		createNCOAuthClient() {
			const url = generateUrl('/apps/integration_openproject/nc-oauth')
			axios.post(url).then((response) => {
				this.state.nc_oauth_client = response.data
				// generate part is complete but still the NC OAuth form is set to
				// edit mode and not completed state so that copy buttons will be available for the user
				this.formMode.ncOauth = F_MODES.EDIT
				this.isFormCompleted.ncOauth = false
			}).catch((error) => {
				showError(
					t('integration_openproject', 'Failed to create Nextcloud OAuth client')
					+ ': ' + error.response.request.responseText
				)
			})
		},
	},
}
</script>

<style scoped lang="scss">
#openproject_prefs {
	div {
		width: 100%;
	}
	.d-flex {
		display: flex;
		align-items: center;
	}
	.pb-1 {
		padding-bottom: .5rem;
	}
	.pb-2 {
		padding-bottom: 1rem;
	}
	.submit-btn {
		background: #397DDA;
		border: #397DDA;
		color: white;

	}
}
</style>
<style lang="scss">
.pencil-icon {
	background-image: url('../../img/pencil.svg');
}

.reset-icon {
	background-image: url('../../img/reset.svg');
}

.check-icon {
	background-image: url('../../img/check.svg');
}

body.theme--dark, body[data-theme-dark], body[data-theme-dark-highcontrast] {
	.pencil-icon, .reset-icon, .copy-icon {
		filter: invert(100%);
	}
}
</style>
