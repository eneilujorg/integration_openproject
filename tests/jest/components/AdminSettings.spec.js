/* jshint esversion: 8 */

import axios from '@nextcloud/axios'
import { createLocalVue, shallowMount } from '@vue/test-utils'
import AdminSettings from '../../../src/components/AdminSettings'
import * as initialState from '@nextcloud/initial-state'

jest.mock('@nextcloud/axios')
jest.mock('@nextcloud/l10n', () => ({
	translate: jest.fn((app, msg) => msg),
	getLanguage: jest.fn(() => ''),
}))
jest.mock('@nextcloud/dialogs', () => ({
	getLanguage: jest.fn(() => ''),
	showError: jest.fn(),
	showSuccess: jest.fn(),
}))

const localVue = createLocalVue()

global.OC = {
	dialogs: {
		confirmDestructive: jest.fn(),
		YES_NO_BUTTONS: 70,
	},
}

const selectors = {
	oauthInstance: '#openproject-oauth-instance',
	oauthClientId: '#openproject-client-id',
	oauthClientSecret: '#openproject-client-secret',
	saveConfigButton: '.save-config-btn',
	updateConfigButton: '.update-config-btn',
}

// eslint-disable-next-line no-import-assign
initialState.loadState = jest.fn(() => {
	return {
		oauth_instance_url: null,
		oauth_client_id: null,
		oauth_client_secret: null,
	}
})

describe('AdminSettings', () => {
	describe('input readonly attribute on focus', () => {
		it.each([
			selectors.oauthClientId,
			selectors.oauthClientSecret,
		])('should not have readonly attribute after the input is focused', async (inputSelector) => {
			const wrapper = getWrapper()
			const inputField = wrapper.find(inputSelector)
			expect(inputField.attributes().readonly).toBeTruthy()

			await inputField.trigger('focus')

			expect(inputField.attributes().readonly).toBeFalsy()

			await inputField.trigger('blur')

			expect(inputField.attributes().readonly).toBeFalsy()
		})
	})
	describe('form submit', () => {
		beforeEach(() => {
			jest.clearAllMocks()
		})
		describe('when the admin config is not complete', () => {
			let wrapper, confirmSpy
			beforeEach(() => {
				confirmSpy = jest.spyOn(global.OC.dialogs, 'confirmDestructive')
				wrapper = getWrapper({
					isAdminConfigOk: false,
				})
			})
			it('should show the save button', async () => {
				expect(wrapper.find(selectors.saveConfigButton)).toMatchSnapshot()
				expect(wrapper.find(selectors.updateConfigButton).exists()).toBeFalsy()
			})
			it('should not trigger confirm dialog on save', async () => {
				const saveConfigButton = wrapper.find(selectors.saveConfigButton)
				const inputField = wrapper.find(selectors.oauthClientId)
				await inputField.setValue('test')
				await saveConfigButton.trigger('click')
				expect(confirmSpy).toBeCalledTimes(0)
			})
		})
		describe('when the admin config status is complete', () => {
			let wrapper, confirmSpy
			beforeEach(() => {
				confirmSpy = jest.spyOn(global.OC.dialogs, 'confirmDestructive')
				wrapper = getWrapper({
					isAdminConfigOk: true,
				})
			})
			it('should show the update button', async () => {
				expect(wrapper.find(selectors.updateConfigButton)).toMatchSnapshot()
				expect(wrapper.find(selectors.saveConfigButton).exists()).toBeFalsy()
			})
			it('should trigger confirm dialog on update', async () => {
				const updateConfigButton = wrapper.find(selectors.updateConfigButton)
				const inputField = wrapper.find(selectors.oauthClientId)
				await inputField.setValue('test')
				await updateConfigButton.trigger('click')

				const expectedDialogMessage = 'Are you sure you want to replace the OpenProject OAuth client details?'
				+ ' Every currently connected user will need to re-authorize this Nextcloud instance to have access to their OpenProject account.'
				const expectedDialogTitle = 'Replace OpenProject OAuth client details'
				const expectedButtonSet = {
					confirm: 'Replace',
					cancel: 'Cancel',
					confirmClasses: 'error',
					type: 70,
				}
				expect(confirmSpy).toBeCalledTimes(1)
				expect(confirmSpy).toHaveBeenCalledWith(
					expectedDialogMessage,
					expectedDialogTitle,
					expectedButtonSet,
					expect.any(Function),
					true
				)
			})
		})
		describe('on saveOptions', () => {
			it.each([
				{ initaialComponentStatus: true, responseStatus: false, finalComponentStatus: false },
				{ initaialComponentStatus: false, responseStatus: true, finalComponentStatus: true },
			])('should update the admin config status as provided by the update response', async ({
				initaialComponentStatus,
				responseStatus,
				finalComponentStatus,
			}) => {
				const axiosSpy = jest.spyOn(axios, 'put')
					.mockImplementationOnce(() => Promise.resolve({
						data: {
							status: responseStatus,
						},
					}))
				const wrapper = getWrapper({
					isAdminConfigOk: initaialComponentStatus,
					state: {
						client_id: 'some-id',
						client_secret: 'some-secret',
						oauth_instance_url: 'some-url',
					},
				})
				await wrapper.vm.saveOptions()
				expect(axiosSpy).toHaveBeenCalledTimes(1)
				expect(wrapper.vm.isAdminConfigOk).toBe(finalComponentStatus)
			})
		})
	})
})

function getWrapper(data = {}) {
	return shallowMount(AdminSettings, {
		localVue,
		mocks: {
			t: (app, msg) => msg,
			generateUrl() {
				return '/'
			},
		},
		data() {
			return {
				...data,
			}
		},
	})
}
