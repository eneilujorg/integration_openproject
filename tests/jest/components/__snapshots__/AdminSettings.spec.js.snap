// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminSettings Nextcloud OAuth values form edit mode should show the form and hide the field values 1`] = `
<div class="nextcloud-oauth-values">
  <formheading-stub index="3" title="Nextcloud OAuth client" iscomplete="true"></formheading-stub>
  <div>
    <textinput-stub value="nc-client-id" id="nextcloud-oauth-client-id" type="text" placeholder="" label="Nextcloud OAuth client ID" hinttext="Copy the following values back into the OpenProject <a class=&quot;link&quot; href=&quot;http://openproject.com/admin/settings/storages&quot; target=&quot;_blank&quot;title=&quot;Administration > File storages&quot;>Administration > File storages</a> as an Administrator." isrequired="true" withcopybtn="true" class="pb-2"></textinput-stub>
    <textinput-stub value="nc-client-secret" id="nextcloud-oauth-client-secret" type="text" placeholder="" label="Nextcloud OAuth client secret" hinttext="Copy the following values back into the OpenProject <a class=&quot;link&quot; href=&quot;http://openproject.com/admin/settings/storages&quot; target=&quot;_blank&quot;title=&quot;Administration > File storages&quot;>Administration > File storages</a> as an Administrator." isrequired="true" withcopybtn="true" class="pb-2"></textinput-stub>
    <button-stub iconclass="check-icon" text="Yes, I have copied these values" class="submit-btn"></button-stub>
  </div>
</div>
`;

exports[`AdminSettings Nextcloud OAuth values form view mode with complete values should show the field values and hide the form 1`] = `
<div class="nextcloud-oauth-values">
  <formheading-stub index="3" title="Nextcloud OAuth client" iscomplete="true"></formheading-stub>
  <div>
    <fieldvalue-stub title="Nextcloud OAuth client ID" value="some-nc-client-id-here" isrequired="true"></fieldvalue-stub>
    <fieldvalue-stub title="Nextcloud OAuth client secret" value="some-nc-client-secret-here" isrequired="true" encryptvalue="true" withinspection="true"></fieldvalue-stub>
    <button-stub iconclass="reset-icon" text="Reset Nextcloud OAuth values"></button-stub>
  </div>
</div>
`;

exports[`AdminSettings OpenProject OAuth values form edit mode should show the form and hide the field values 1`] = `
<div class="openproject-oauth-values">
  <div class="form-heading">
    <div class="index">
      2
    </div>
    <div class="title">
      OpenProject OAuth settings
    </div>
  </div>
  <div>
    <div class="text-input-wrapper pb-2">
      <div class="text-input">
        <div class="text-input-label">
          OpenProject OAuth client ID *
        </div> <input id="openproject-oauth-client-id" type="text" placeholder="" class="full-width">
        <div class="text-input-messages">
          <div class="text-input-hint">Go to your OpenProject <a class="link" href="http://openproject.com/admin/settings/storages" target="_blank" title="Administration > File storages">Administration &gt; File storages</a> as an Administrator and start the setup and copy the values here.</div>
        </div>
      </div>
      <!---->
    </div>
    <div class="text-input-wrapper pb-2">
      <div class="text-input">
        <div class="text-input-label">
          OpenProject OAuth client secret *
        </div> <input id="openproject-oauth-client-secret" type="text" placeholder="" class="full-width">
        <div class="text-input-messages">
          <div class="text-input-hint">Go to your OpenProject <a class="link" href="http://openproject.com/admin/settings/storages" target="_blank" title="Administration > File storages">Administration &gt; File storages</a> as an Administrator and start the setup and copy the values here.</div>
        </div>
      </div>
      <!---->
    </div> <button disabled="disabled" class="submit-btn b-disabled" data-test-id="submit-op-oauth-btn">
      <div class="b-icon check-icon"></div> <span>Save</span>
      <!---->
    </button>
  </div>
</div>
`;

exports[`AdminSettings OpenProject OAuth values form view mode and completed state should show field values and hide the form if server host form is complete 1`] = `
<div class="openproject-oauth-values">
  <div class="form-heading">
    <div class="complete">
      <div class="icon icon-checkmark"></div>
    </div>
    <div class="title green-text">
      OpenProject OAuth settings
    </div>
  </div>
  <div>
    <div class="field-value"><b class="title">
        OpenProject OAuth client ID*:
      </b>
      &nbsp;
      <span>openproject-client-id</span>
      <!---->
    </div>
    <div class="field-value pb-1"><b class="title">
        OpenProject OAuth client secret*:
      </b>
      &nbsp;
      <div data-test-id="encrypted-value"><span>openproj***************</span></div>
      <!---->
    </div> <button class="" data-test-id="reset-op-oauth-btn">
      <div class="b-icon reset-icon"></div> <span>Reset OpenProject OAuth values</span>
      <!---->
    </button>
  </div>
</div>
`;

exports[`AdminSettings server host url form edit mode cancel button should be visible when the form is in completed state with edit mode 1`] = `
<button class="" data-test-id="cancel-edit-server-host-btn">
  <!----> <span>Cancel</span>
  <!---->
</button>
`;

exports[`AdminSettings server host url form edit mode submit button should set the input to error state when the url is invalid when clicked 1`] = `
<div class="text-input-wrapper pb-2">
  <div class="text-input">
    <div class="text-input-label">
      OpenProject host *
    </div> <input id="openproject-oauth-instance" type="text" placeholder="https://www.my-openproject.com" class="full-width error">
    <div class="text-input-messages">
      <div class="text-input-error">
        Please introduce a valid OpenProject host name
      </div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`AdminSettings server host url form view mode and completed state reset button should be visible when the form is in completed state 1`] = `
<button class="" data-test-id="reset-server-host-btn">
  <div class="b-icon pencil-icon"></div> <span>Edit server information</span>
  <!---->
</button>
`;

exports[`AdminSettings server host url form view mode and completed state should show field value and hide the input field 1`] = `
<div class="openproject-server-host">
  <div class="form-heading">
    <div class="complete">
      <div class="icon icon-checkmark"></div>
    </div>
    <div class="title green-text">
      OpenProject Server
    </div>
  </div>
  <div class="field-value pb-1"><b class="title">
      OpenProject host*:
    </b>
    &nbsp;
    <span>http://openproject.com</span>
    <!---->
  </div>
  <div class="d-flex"><button class="" data-test-id="reset-server-host-btn">
      <div class="b-icon pencil-icon"></div> <span>Edit server information</span>
      <!---->
    </button>
    <!---->
    <!---->
  </div>
</div>
`;
